dot/services_test.go

```diff
@@ -0,0 +1,639 @@
+		cfg       *Config
+		stateSrvc *state.Service
+	}
+	tests := []struct {
+		name string
+		args args
+		want *rpc.HTTPServer
+		err  error
+	}{
+		{
+			name: "base state",
+			args: args{
+				cfg:       cfg,
+				stateSrvc: stateSrvc,
+			},
+			want: &rpc.HTTPServer{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			no := nodeBuilder{}
+			rpcParams := rpcServiceSettings{
+				config: tt.args.cfg,
+				state:  tt.args.stateSrvc,
+			}
+			got, err := no.createRPCService(rpcParams)
+			assert.ErrorIs(t, err, tt.err)
+
+			if tt.want != nil {
+				assert.NotNil(t, got)
+			} else {
+				assert.Nil(t, got)
+			}
+		})
+	}
+}
+
+func Test_nodeBuilder_createGRANDPAService(t *testing.T) {
+	t.Parallel()
+	cfg := NewTestConfig(t)
+	stateSrvc := newStateService(t)
+	ks := keystore.NewGlobalKeystore()
+	kr, err := keystore.NewEd25519Keyring()
+	require.NoError(t, err)
+	ks.Gran.Insert(kr.Alice())
+
+	networkConfig := &network.Config{
+		BasePath:   t.TempDir(),
+		BlockState: stateSrvc.Block,
+		RandSeed:   2,
+	}
+	networkSrvc, err := network.NewService(networkConfig)
+	require.NoError(t, err)
+	type args struct {
+		cfg *Config
+		st  *state.Service
+		ks  keystore.Keystore
+		net *network.Service
+	}
+	tests := []struct {
+		name string
+		args args
+		want *grandpa.Service
+		err  error
+	}{
+		{
+			name: "base case",
+			args: args{
+				cfg: cfg,
+				st:  stateSrvc,
+				ks:  ks.Gran,
+				net: networkSrvc,
+			},
+			want: &grandpa.Service{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			no := nodeBuilder{}
+			got, err := no.createGRANDPAService(tt.args.cfg, tt.args.st, nil, tt.args.ks, tt.args.net,
+				nil)
+			assert.ErrorIs(t, err, tt.err)
+			if tt.want != nil {
+				assert.NotNil(t, got)
+			} else {
+				assert.Nil(t, got)
+			}
+		})
+	}
+}
+
+func Test_createRuntime(t *testing.T) {
+	t.Parallel()
+	cfg := NewTestConfig(t)
+	stateSrvc := newStateService(t)
+	code, err := stateSrvc.Storage.LoadCode(nil)
+	require.NoError(t, err)
+
+	cfgLife := NewTestConfig(t)
+	cfgLife.Core.WasmInterpreter = life.Name
+
+	type args struct {
+		cfg  *Config
+		ns   runtime.NodeStorage
+		st   *state.Service
+		code []byte
+	}
+	tests := []struct {
+		name string
+		args args
+		want runtime.Instance
+		err  error
+	}{
+		{
+			name: "wasmer runtime",
+			args: args{
+				cfg:  cfg,
+				ns:   runtime.NodeStorage{},
+				st:   stateSrvc,
+				code: code,
+			},
+			want: &wasmer.Instance{},
+			err:  nil,
+		},
+		{
+			name: "wasmer life",
+			args: args{
+				cfg:  cfgLife,
+				ns:   runtime.NodeStorage{},
+				st:   stateSrvc,
+				code: code,
+			},
+			want: &life.Instance{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := createRuntime(tt.args.cfg, tt.args.ns, tt.args.st, nil, nil, tt.args.code)
+			assert.ErrorIs(t, err, tt.err)
+			if tt.want != nil {
+				assert.NotNil(t, got)
```

Comment:
If you don't want to inject interfaces for the `wasmer.Instance` and `life.Instance` constructors, at least make sure that the types are what you're expecting.

Reply:
Added check for instance type.